text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,621 @@
+  this can be represented in the graph using edges. This means that
+  incremental compilation can also work across crates.
+- **IR nodes.** Represent some portion of the computed IR. For
+  example, the MIR representation of a fn body, or the `ty`
+  representation of a fn signature. These also frequently correspond
+  to a single entry in one of the various compiler hashmaps. These are
+  the outputs (and intermediate steps) of the compilation process
+  - Examples:
+    - `ITEM_TYPE(X)` -- entry in the obscurely named `tcache` table
+      for `X` (what is returned by the rather-more-clearly-named
+      `lookup_item_type`)
+    - `PREDICATES(X)` -- entry in the `predicates` table
+    - `ADT(X)` -- ADT node for a struct (this may want to be more
+      fine-grained, particularly to cover the ivars)
+    - `MIR(X)` -- the MIR for the item `X`
+    - `LLVM(X)` -- the LLVM IR for the item `X`
+    - `OBJECT(X)` -- the object code generated by compiling some item
+      `X`; the precise way that this is saved will depend on whether
+      we use `.o` files that are linked together, or if we attempt to
+      amend the shared library in place.
+- **Procedure nodes.** These represent various passes performed by the
+  compiler. For example, the act of type checking a fn body, or the
+  act of constructing MIR for a fn body. These are the "glue" nodes
+  that wind up reading the inputs and creating the outputs, and hence
+  which ultimately tie the graph together.
+  - Examples:
+    - `COLLECT(X)` -- the collect code executing on item `X`
+    - `WFCHECK(X)` -- the wfcheck code executing on item `X`
+    - `BORROWCK(X)` -- the borrowck code executing on item `X`
+
+To see how this all fits together, let's consider the graph for a
+simple example:
+
+```rust
+fn foo() {
+    bar();
+}
+
+fn bar() {
+}
+```
+
+This might generate a graph like the following (the following sections
+will describe how this graph is constructed). Note that this is not a
+complete graph, it only shows the data needed to produce `MIR(foo)`.
+
+```
+BODY(foo) ----------------------------> TYPECK(foo) --> MIR(foo)
+                                          ^ ^ ^ ^         |
+SIG(foo) ----> COLLECT(foo)               | | | |         |
+                 |                        | | | |         v
+                 +--> ITEM_TYPE(foo) -----+ | | |      LLVM(foo)
+                 +--> PREDICATES(foo) ------+ | |         |
+                                              | |         |
+SIG(bar) ----> COLLECT(bar)                   | |         v
+                 |                            | |     OBJECT(foo)
+                 +--> ITEM_TYPE(bar) ---------+ |
+                 +--> PREDICATES(bar) ----------+
+```
+
+As you can see, this graph indicates that if the signature of either
+function changes, we will need to rebuild the MIR for `foo`. But there
+is no path from the body of `bar` to the MIR for foo, so changes there
+need not trigger a rebuild.
```

Comment:
I believe this is an understood limitation, but it may be worth pointing out again that this wouldn't allow `bar` to be inlined into `foo`, otherwise you'd end up with two versions of `bar`.